### DISCUSSION TITLE
React.ChangeEventHandler<HTMLInputElement> 와 React.ChangeEvent<HTMLInputElement>의 차이

### DISCUSSION THREAD

**Discussion by Arsture** [wafflestudio/seminar-2021#628] [Q&A] [answered]
onChange 속성에 커서를 가져다 대면 아래와 같이 뜹니다.
![image](https://user-images.githubusercontent.com/87017787/140261016-bf7daab6-58f9-4000-b71c-4c9390d26673.png)
그래서 아래 함수 파라미터 e의 type을 지정할 때 React.ChangeEventHandler<HTMLInputElement>로 지정하면 아래와 같은 오류가 뜹니다.
![image](https://user-images.githubusercontent.com/87017787/140261073-d5308c43-1963-4887-879f-8c1413eb6e72.png)
그렇지만 e의 type을 React.ChangeEvent<HTMLInputElement>로 지정하면 오류없이 잘 되는데, 어떤 차이가 있는 건가요??

**Comment by woohm402** [comment, marked as answer]
문장 그대로 해석해 주시면 됩니다. `ChangeEventHandler` 는 **이벤트 핸들러** (함수) 의 타입이고, `ChangeEvent`는 **이벤트** 의 타입입니다.

`ChangeEvent`를 이용해서 타입을 잡으시겠다면
```ts
const onChange = (e: React.ChangeEvent<HTMLInputElement>) => { ... };
```

해주시면 되고, ChangeEventHandler를 이용하시겠다면
```ts
const onChange: React.ChangeEventHandler = (e) => { ... };
```

해주시면 됩니다. 두 경우 모두 `e`의 타입이 `ChangeEvent` 로 잘 잡힙니다.

**Comment by woohm402** [comment]
아주 조금만 더 딥하게 파볼게요 (리액트는 오픈소스니까요 👍 )

`@types/react` 가 정의되어 있는 [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped) 레포를 파 보면서 차이점을 뜯어보겠습니다.

1. `ChangeEventHandler<T>`는 특별한 게 아니고, 그냥 `EventHandler<ChangeEvent<T>>`의 type alias (별칭)입니다. ([해당 코드](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/333a18ceae46e956fe61bc4440206933079ad624/types/react/index.d.ts#L1316))

2. `EventHandler<T>`는 `{ bivarianceHack(event: T): void }["bivarianceHack"]` 의 type alias 입니다. ([해당 코드](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/333a18ceae46e956fe61bc4440206933079ad624/types/react/index.d.ts#L1307))

3. `bivarianceHack`은 특별한 친구가 아닙니다.
    - 이름이 저래서 뭐 있을 것 같지만, `{ bivarianceHack(event: T): void }["bivarianceHack"]` 은 `{ temp(event: T): void }["temp"]`와 동일합니다. 타입이라 헷갈릴 수 있으니 값을 가지고 일반적인 예시를 들자면 `const val = { a: 1 }['a']; // 1` 뭐 이런 느낌입니다. 그러니까 `{ bivarianceHack(event: T): void }["bivarianceHack"]` 은 `(event: T) => void` 와 동일한 형식입니다.
    - 근데 왜 굳이 `(event: T) => void` 라는 친숙한 타입을 안 쓰고 `bivarianceHack`이라는 어려운 말을 써서 한 단계 감쌌냐: 이건 타스의 [뭔가](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-6.html#strict-function-types) 를 피하기 위한 하나의 꼼수로 봐주시면 됩니다. 여기서 더 나가면 복잡해지고 저도 정확힌 모르니 패스하겠습니다. 자세한 내용은 [여기](https://stackoverflow.com/questions/52667959/what-is-the-purpose-of-bivariancehack-in-typescript-types) 잘 설명되어 있는 것 같습니다.

자 여기까지 공부한 걸 종합해 보겠습니다.

```ts
ChangeEventHandler<HTMLInputElement>
== EventHandler<ChangeEvent<HTMLInputElement>>
== (e: ChangeEvent<HTMLInputElement>) => void
```

결론은 일반적인 상황에서는 `ChangeEventHandler` 랑 `(e: ChangeEvent) =>` 중 뭘 쓰시든 똑같습니다. 둘 차이는 대충 선호 차이인 것 같고.. 저는 핸들러까지 명시하는 걸 선호하긴 합니다

**Reply by Arsture** [reply]
아하,,, 감사합니당!!
